src/GraphBLAS-sharp.Backend/Matrix/CSRMatrix/Kronecker.fs

```diff
@@ -0,0 +1,376 @@
+                                    newMatrix
+                                | _ -> newMatrix
+
+                    // inserting new matrix resulting from multiplication by non None element of left matrix
+                    let columnOffset = currentColumn * matrixRight.ColumnCount
+
+                    let operand =
+                        Some leftMatrixVals.[firstElementIndex + offset]
+                        |> clContext.CreateClCell
+
+                    let mappedMatrix =
+                        mapWithValueToCOO queue allocationMode operand matrixRight
+
+                    queue.Post(Msg.CreateFreeMsg<_>(operand))
+
+                    let newMatrix =
+                        insert queue allocationMode rowOffset columnOffset mappedMatrix resultMatrix
+
+                    resultMatrix <-
+                        match resultMatrix, mappedMatrix with
+                        | Some oldMatrix, Some m ->
+                            queue.Post(Msg.CreateFreeMsg<_>(oldMatrix.Rows))
+                            queue.Post(Msg.CreateFreeMsg<_>(oldMatrix.Columns))
+                            queue.Post(Msg.CreateFreeMsg<_>(oldMatrix.Values))
+                            queue.Post(Msg.CreateFreeMsg<_>(m.Rows))
+                            queue.Post(Msg.CreateFreeMsg<_>(m.Columns))
+                            queue.Post(Msg.CreateFreeMsg<_>(m.Values))
+                            newMatrix
+                        | _ -> newMatrix
+
+                // inserting matrices resulting from multiplication by 0 to the rest of the place in row
+                let startColumn =
+                    match NNZInRow with
+                    | 0 -> 0
+                    | _ ->
+                        leftMatrixCols.[firstElementIndex + NNZInRow - 1]
+                        + 1
+
+                for i in startColumn .. matrixLeft.ColumnCount - 1 do
+                    let columnOffset = i * matrixRight.ColumnCount
+
+                    let newMatrix =
+                        insert queue allocationMode rowOffset columnOffset mapWithZero resultMatrix
+
+                    resultMatrix <-
+                        match resultMatrix, mapWithZero with
+                        | Some oldMatrix, Some _ ->
+                            queue.Post(Msg.CreateFreeMsg<_>(oldMatrix.Rows))
+                            queue.Post(Msg.CreateFreeMsg<_>(oldMatrix.Columns))
+                            queue.Post(Msg.CreateFreeMsg<_>(oldMatrix.Values))
+                            newMatrix
+                        | _ -> newMatrix
+
+            match resultMatrix with
+            | None -> None
+            | Some m ->
+                Some
+                    { m with
+                          RowCount = matrixLeft.RowCount * matrixRight.RowCount
+                          ColumnCount = matrixLeft.ColumnCount * matrixRight.ColumnCount }
+
+    let run<'a, 'b, 'c when 'a: struct and 'b: struct and 'c: struct and 'c: equality>
```

Comment:
Why is it necessary to explicitly pair the conversion function? It seems that at the same time we lose in modularity